odoo/addons/base/transpiler/sourcemap_generator.py

```diff
@@ -0,0 +1,168 @@
+
+            if (source, line) not in self.cache:
+                self.cache[(source, line)] = "".join([encoded_column, base64vlq_encode(source), base64vlq_encode(line), encoded_column])
+            result += self.cache[source, line]
+        return result
+
+    def to_json(self):
+        """
+        Generates the json sourcemap.
+        It is the main function that assembles all the pieces.
+
+        :return {str} valid sourcemap in json format
+        """
+        mapping = {
+            "version": self._version,
+            "sources": list(self._sources.keys()),
+            "mappings": self._serialize_mappings(),
+            "sourcesContent": [self._sourcesContents[source] for source in self._sources]
+        }
+        if (self._file is not None):
+            mapping["file"] = self._file
+
+        if (self._sourceRoot is not None):
+            mapping["sourceRoot"] = self._sourceRoot
+
+        return mapping
+
+    def add_source(self, source_name, source_content, previous_generated_line, nb_skip_line_output=0):
+        """Adds a new source file in the sourcemap. All the lines of the source file will be mapped line by line
+        to the generated file from the (previous_generated_line + nb_skip_line_output). All lines between
+        previous_generated_line and (previous_generated_line + nb_skip_line_output) will
+        be mapped to line 1 of the source file.
+
+        Example:
+            ls 1 = Line 1 from new source file
+            lg 1 = Line 1 from genereted file
+            ls 1 <=> lg 1 Line 1 from new source file is map to  Line 1 from genereted file
+            nb_ls = number of lines in the new source file
+
+            Step 1:
+            ls 1 <=> lg previous_generated_line + 1
+
+            Step 2:
+            ls 1 <=> lg previous_generated_line + nb_skip_line_output + 1
+            ls 2 <=> lg previous_generated_line + nb_skip_line_output + 2
+            ...
+            ls nb_ls <=> lg previous_generated_line + nb_skip_line_output + nb_ls
+
+
+        :param source_name: name of the source to add
+        :param source_content: content of the source to add
+        :param previous_generated_line: Line where we start to map the new source
+        :param nb_skip_line_output: Number of lines to pass in the generated file before starting mapping line by line
+
+        :return number of lines added
+        """
+        source_line_count = len(source_content.split("\n"))
+
+        if (source_name not in self._sources):
+            self._sources[source_name] = len(self._sources)
+
+        self._sourcesContents[source_name] = source_content
+        if nb_skip_line_output > 0:
+            # adds a mapping between the first line of the source
+            # and the first line of the corresponding code in the generated file.
+            self._mappings.append({
+                "generatedLine": previous_generated_line + 1,
+                "originalLine": 1,
```

Comment:
Since the mapping serialisation adjusts lines back to 0-indexed, maybe this could store them zero-indexed in the first place?